consensus/view_change.go

```diff
@@ -214,228 +200,47 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
-				copy(preparedMsg.Payload[:], recvMsg.Payload[32:])
-				preparedMsg.SenderPubkey = newLeaderKey
-				consensus.getLogger().Info().Msg("[onViewChange] New Leader Prepared Message Added")
-				consensus.FBFTLog.AddMessage(&preparedMsg)
-
-				consensus.FBFTLog.AddBlock(preparedBlock)
-			}
-		}
-		consensus.getLogger().Info().
-			Str("validatorPubKey", senderKey.Bytes.Hex()).
-			Msg("[onViewChange] Add M1 (prepared) type message")
-		consensus.bhpSigs[recvMsg.ViewID][senderKey.Bytes.Hex()] = recvMsg.ViewchangeSig
-		consensus.bhpBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true) // Set the bitmap indicating that this validator signed.
-	}
-
-	// check and add viewID (m3 type) message signature
-	if _, ok := consensus.viewIDSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]; ok {
-		consensus.getLogger().Debug().
-			Str("validatorPubKey", senderKey.Bytes.Hex()).
-			Msg("[onViewChange] Already Received M3(ViewID) message from the validator")
+	members := consensus.Decider.Participants()
+	consensus.vc.AddViewIDKeyIfNotExist(recvMsg.ViewID, members)
+
+	// do it once only per viewID/Leader
+	if err := consensus.vc.InitPayload(consensus.FBFTLog,
+		recvMsg.ViewID,
+		recvMsg.BlockNum,
+		consensus.priKey); err != nil {
+		consensus.getLogger().Error().Err(err).Msg("Init Payload Error")
 		return
 	}
-	viewIDHash := make([]byte, 8)
-	binary.LittleEndian.PutUint64(viewIDHash, recvMsg.ViewID)
-	if !recvMsg.ViewidSig.VerifyHash(recvMsg.SenderPubkey.Object, viewIDHash) {
-		consensus.getLogger().Warn().
-			Uint64("MsgViewID", recvMsg.ViewID).
-			Msg("[onViewChange] Failed to Verify M3 Message Signature")
+
+	err = consensus.vc.ProcessViewChangeMsg(consensus.FBFTLog, consensus.Decider, recvMsg)
+	if err != nil {
+		consensus.getLogger().Error().Err(err).
+			Uint64("viewID", recvMsg.ViewID).
+			Uint64("blockNum", recvMsg.BlockNum).
+			Str("msgSender", recvMsg.SenderPubkey.Bytes.Hex()).
+			Msg("Verify View Change Message Error")
 		return
 	}
-	consensus.getLogger().Info().
-		Str("validatorPubKey", senderKey.Bytes.Hex()).
-		Msg("[onViewChange] Add M3 (ViewID) type message")
-
-	consensus.viewIDSigs[recvMsg.ViewID][senderKey.Bytes.Hex()] = recvMsg.ViewidSig
-	// Set the bitmap indicating that this validator signed.
-	consensus.viewIDBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true)
-	consensus.getLogger().Info().
-		Int("have", len(consensus.viewIDSigs[recvMsg.ViewID])).
-		Int64("total", consensus.Decider.ParticipantsCount()).
-		Msg("[onViewChange]")
 
 	// received enough view change messages, change state to normal consensus
-	if consensus.Decider.IsQuorumAchievedByMask(consensus.viewIDBitmap[recvMsg.ViewID]) {
+	if consensus.Decider.IsQuorumAchievedByMask(consensus.vc.GetViewIDBitmap(recvMsg.ViewID)) {
+		consensus.getLogger().Info().Msg("[onViewChange] View Change Message Quorum Reached")
 		consensus.current.SetMode(Normal)
 		consensus.LeaderPubKey = newLeaderKey
```

Comment:
These state changes have high risk for race conditions. As when multiple viewChangeMsg are received that all makes IsQuorumAchievedByMask() == true, then these states will be updated concurrently. And this is very likely to happen since those messages can arrive very closely.

Reply:
I think it's better to keep the old lock logic to avoid troubles. The view change protocol is not performance concerned after all, stability of the nodes is much more important for view change.

Reply:
why is there an issue if multiple view change messages arrived if they are all valid? we can check the mode during the quorum check as well.